core/domain/exp_domain.py

```diff
@@ -2514,12 +2515,1063 @@ def _convert_states_v50_dict_to_v51_dict(
+    def _update_general_state(
+        cls, states_dict: Dict[str, state_domain.StateDict]
+    ) -> Dict[str, state_domain.StateDict]:
+        """Handles errored data for the general exploration state, performs the
+        following -
+            - If destination is `try again` and the value of labeled_as_correct
+            is True, replaces it with False
+            - If refresher_exploration_id is not None for lesson, marks it as
+            None
+
+        Args:
+            states_dict: dict. A dict where each key-value pair represents,
+                respectively, a state name and a dict used to initialize a
+                State domain object.
+
+        Returns:
+            dict. The converted states_dict.
+        """
+        for state_name, state_dict in states_dict.items():
+            answer_groups = state_dict['interaction']['answer_groups']
+            for answer_group in answer_groups:
+                # lab_as_correct should not be True if dest is try again.
+                if (
+                    answer_group['outcome']['dest'] == state_name and
+                    answer_group['outcome']['labelled_as_correct']
+                ):
+                    answer_group['outcome']['labelled_as_correct'] = False
+
+                # refresher_exploration_id be None for all lessons.
+                if (
+                    answer_group['outcome']['refresher_exploration_id'] is
+                    not None
+                ):
+                    answer_group['outcome']['refresher_exploration_id'] = None
+
+            # refresher_exploration_id be None for all lessons(def outcome).
+            default_outcome = state_dict['interaction']['default_outcome']
+            if default_outcome is not None:
+                def_ref_exp_id = default_outcome['refresher_exploration_id']
+                if def_ref_exp_id is not None:
+                    state_dict['interaction']['default_outcome'][
+                        'refresher_exploration_id'] = None
+
+            state_dict['interaction']['answer_groups'] = answer_groups
+        return states_dict
+
+    @classmethod
+    def _choices_should_be_unique_and_non_empty(
+        cls, choices: List[str],
+        answer_groups: List[state_domain.AnswerGroup]
+    ) -> Tuple(Any):
+        """Handles choices present in the ItemSelectionInput or
+        in MultipleChoiceInput interactions, implements the following -
+            - If only one choice is empty then simply removes it
+            - If multiple choices are empty replace them with `Choice 1` ,
+            `Choice 2` etc
+            - If choices are duplicate, removes the later choice
+
+        Args:
+            choices: List[str]. A list of choices.
+
+        Returns:
+            choices_to_save: List[str]. The list of valid choices.
+        """
+        # Misbehaving when after removing invalid choices and the choices count becomes below 4.
+        empty_choices = []
+        seen_choices = []
```

Comment:
what's the point of the `seen_choices` variable?

Reply:
`seen_choices` will help us to detect duplicates, if the choice is present in the `seen_choices` then it is a duplicate and we can remove it.